Added option to set max ALSA volume in dB

JWahle opened the pull request:
Fixes #275 

I added a field to set the volume in dB.
If one is set, the other is updated accordingly.
If setting dB fails for some reason, the % value stays and dB value becomes blank.

![set_volume_in_dB](https://user-images.githubusercontent.com/7657699/104079160-e556c700-5221-11eb-9df3-1651b88404dd.png)

I also thought about adding some indicator, which clarifies whether % or dB value is used to set the volume.
Since setting the % usually gives a fractional dB value, it should be obvious, which one is currently in use.
However, if you know a simple, pretty solution to this, I will implement it.
I tried with simple checkboxes, but didn't like the look of it:
![set_volume_in_dB_with_checkbox](https://user-images.githubusercontent.com/7657699/104080253-2ac9c300-5227-11eb-98ad-3ab5fdf809dd.png)

What do you think?

moodeaudio commented:
Whats the usage scenario for having two different volume scales for Max ALSA vol ?

The only dB value that works in amixer sset command is "0dB". It won't accept negative values and so it's sort of useless as a general method for setting volume. Also using a dB scale would require knowing what the the audio device's hardware volume scale was so the UI volume knob could be scaled correctly.

JWahle commented:
> Whats the usage scenario for having two different volume scales for Max ALSA vol ?

The usage scenario is the one from #275:
You cannot set exactly 0dB on the Pi's HDMI output using percentages (96% => -0.25dB)

> The only dB value that works in amixer sset command is "0dB"

It is possible to set negative values as [explained here](https://github.com/moode-player/moode/issues/275#issuecomment-747123899)

> Also using a dB scale would require knowing what the the audio device's hardware volume scale was so the UI volume knob could be scaled correctly.

I don't think this is relevant. This PR does not change the scale of the volume knob. The "Max ALSA volume" setting is (and was) only used, when MPD volume is set to software or disabled.

JWahle commented:
There is a different approach for the UI that I want to try out before merging.

moodeaudio commented:
Ok the -- means "no more options" would allow -dB values.

So this is to remedy the 100% = +4dB for the HDMI and Headphone ports?

The challenge is that adding this complicates the Max ALSA setting for the majority of users that prolly won't need it. Another approach would be to test in all the files for ((HDMI or Headphone) and (Software or Disabled)) and then use 0dB instead of 100%..

JWahle commented:
![set_volume_in_db_with_select](https://user-images.githubusercontent.com/7657699/104081780-3de19080-5231-11eb-8006-c3cb3dd2a381.png)

What do you think of this UI?
Setting 100% is as easy as before and one can even detect the 4dB error with HDMI output.
This is not possible with my first proposal.

moodeaudio commented:
I would leave the UI alone and do the detection under the covers so user just sets 100% in the one field and the code determines whether to do 100% or 0dB. Which one is used could be explained in the help text.

JWahle commented:
That would also be fine with me - and probably prevents someone else from having this problem.
So, what logic would you use?
Currently, on the HDMI port the volumes are mapped: 100% = 4.00dB, 99% = 2.94dB, 98% = 1.88dB, 97% = 0.81dB, 96% = -0.25dB
If only 100% is set to 0dB, then 99% is louder than 100%.
Or should no volume above 0dB be possible? (100% = 99% = 98% = 97% = 0dB)

moodeaudio commented:
I think a test in all the files that set 0dB for ( (HDMI or Headphone) and (Software or Disabled) ) and then use "0dB" instead of "100%" in the amixer command.

It's an edge case though cos most users are into HAT or USB Audio devices and don't use the on-board HDMI/Headphone ports. It might be better to lobby the Pi Foundation to ditch to +4dB boost or at least make it optional.

monkeypox commented:
Instead of two input fields why not just use one and then compute and show the other value for people that want that number, e.g. for 100% show 0dB or vice versa if you think dB is the more important thing to show. 

Sent from my iPhone

> On Jan 8, 2021, at 6:03 PM, JWahle <notifications@github.com> wrote:
> 
> ﻿
> I added a field to set the volume in dB.
> If one is set, the other is updated accordingly.
> If setting dB fails for some reason, the % value stays and dB value becomes blank.
> 
> 
> 
> I also thought about adding some indicator, which clarifies whether % or dB value is used to set the volume.
> Since setting the % usually gives a fractional dB value, it should be obvious, which one is currently in use.
> However, if you know a simple, pretty solution to this, I will implement it.
> I tried with simple checkboxes, but didn't like the look of it:
> 
> 
> What do you think?
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/moode-player/moode/pull/298
> 
> Commit Summary
> 
> Added option to set max ALSA volume in dB
> File Changes
> 
> M var/local/www/commandw/slpower.sh (14)
> M var/local/www/commandw/spotevent.sh (16)
> M var/local/www/commandw/spspost.sh (14)
> M var/local/www/commandw/spspre.sh (6)
> M var/local/www/db/moode-sqlite3.db.sql (2)
> M www/command/util.sh (42)
> M www/command/worker.php (13)
> M www/inc/playerlib.php (2)
> M www/snd-config.php (47)
> M www/templates/snd-config.html (5)
> Patch Links:
> 
> https://github.com/moode-player/moode/pull/298.patch
> https://github.com/moode-player/moode/pull/298.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

moodeaudio commented:
Its not really about showing a particular scale for that setting. The on-board HDMI and Headphone ports are configured by RaspiOS/Kernel to be ALSA 100% = +4dB instead of = 0dB. Typically one would just back off ALSA max volume to 96% or something like that to get to 0dB but the percentage scale doesn't allow fractions and none of the whole number values yield exactly 0dB.

Apparently though there is a way to get exact 0dB by using dB scale in the amixer command.
```
amixer sset Digital 0dB
```

This sets HDMI or Headphone port to exactly 0dB. If the -- "no more arguments" switch is used the command will accept negative dB values which in theory would allow using a pure dB scale for example 0 to -127dB. This would be ideal but prolly a fair amount of work to reflect through the UI and knob especially if the device's dB range had to be considered for example some devices only go to -96dB or -115dB
```
amixer -- sset Master -82dB
```

monkeypox commented:
Reading this on GitHub to get context I agree with Tim about just doing the right thing behind the scenes wrt 0 dB and filing an issue with the Pi foundation about the +4 dB issue with headphones and the hdmi port.

JWahle commented:
I reverted to the old UI and added the info text.
`util.sh` now contains an option `set-alsavol-to-max`, which implements the workaround.
It is only effective through the Max ALSA Volume setting - not when moOde volume control is used to change ALSA volume (MPD Settings: Mixer Type = "Hardware").
I also refactored all placed that called `util.sh set-alsavol ...` to use `set-alsavol-to-max`.
If you are concerned regarding the `set-alsavol` refactoring, or want the fix to apply with `Mixer Type = Hardware`, I can also fix that.

JWahle commented:
I just tested this with the headphone output, but it is not affected. 100% or 4dB output volume does not distort, so I excluded it from this patch.

moodeaudio commented:
Since ALSA volume is set by util.sh I think that might be the only file that needs to be modified. The code block that sets ALSA volume should do the test below. It just need to get some sql values. This way we can avoid another setting on System config and just explain in the help test what happens when the mixer is HDMI or Headphones.

```
if $3 == 100 and (HDMI or Headphone) and (Software or Disabled) then "0dB" else "100%"
```

```
pi@rp2:~ $ moodeutl -q "select * from cfg_system where param='amixname'"
36|amixname|HDMI

pi@rp2:~ $ moodeutl -q "select * from cfg_mpd where param='mixer_type'" 
22|mixer_type|hardware
```

JWahle commented:
With "I reverted to the old UI" I meant, that I reverted to the vanilla moOde UI and did pretty much what you said.
There is no additional setting in the system config.
I rebased my commits, to remove some unnecessary changes from the diff.
This is what I changed in the individual files:
File | Changes
----|----
snd-config.html | Added the info, that 100% on HDMI is replaced with 0dB and other values are unchanged.
snd-config.php | Some variable names for the template were wrong - I fixed that.
util.sh | Implemented the suggested changes through the parameter `set-alsavol-to-max`
other *.sh files | Every file contained duplicated code, that loaded `alsavol_max` and `amixname` from the db<br>and call `util.sh amixname alsavol_max`.<br>I replaced all that with just `util.sh set-alsavol-to-max`.
other .php files | I replaced these calls:<br>`sysCmd('/var/www/command/util.sh set-alsavol ' . '"' . $_SESSION['amixname']  . '" ' . $_SESSION['alsavolume_max']);`<br>with `sysCmd('/var/www/command/util.sh set-alsavol-to-max');`

moodeaudio commented:
I'm still not following the changes, especially the introduction of get/set-alsavolume-db in util.sh. I though the only concern was 100% causes +4dB on HDMI or Headphone which can be remedied by using "0dB" instead of "100%" in the amixer command. I think this can be addressed by some mods to just util.sh.

moodeaudio commented:
I just committed an update that addresses the HDM/Headphone 0dB issue.
https://github.com/moode-player/moode/commit/fc894beccf9cc58bea10d0470a87c6bddbd1fb01

Please review. If this takes of the issue then go ahead and close this PR. If you would like to propose a general method of using dB for ALSA volume then feel free to open an issue so we can discuss.

JWahle commented:
With one small exception, the commit seems fine to me: +4dB on Headphone output is ok and does not distort.
See [here](https://github.com/moode-player/moode/commit/fc894beccf9cc58bea10d0470a87c6bddbd1fb01#r45892700)

moodeaudio commented:
Ok got it. I'll push an update in a bit.

JWahle commented:
I tested this and it works as I expected, thanks.
